src/ExplainProvenanceSLD.h

```diff
@@ -0,0 +1,231 @@
+                tuple >> levelNum;
+
+                return std::make_pair(ruleNum, levelNum);
+            }
+        }
+
+        // if no tuple exists
+        return std::make_pair(-1, -1);
+    }
+
+public:
+    ExplainProvenanceSLD(SouffleProgram& prog) : ExplainProvenance(prog) {
+        setup();
+    }
+
+    void setup() override {
+        // for each clause, store a mapping from the head relation name to body relation names
+        for (auto& rel : prog.getAllRelations()) {
+            std::string name = rel->getName();
+
+            // only process info relations
+            if (name.find("@info") == std::string::npos) {
+                continue;
+            }
+
+            // find all the info tuples
+            for (auto& tuple : *rel) {
+                std::vector<std::string> bodyRels;
+
+                RamDomain ruleNum;
+                tuple >> ruleNum;
+
+                for (size_t i = 1; i < rel->getArity() - 1; i++) {
+                    std::string bodyRel;
+                    tuple >> bodyRel;
+                    bodyRels.push_back(bodyRel);
+                }
+
+                std::string rule;
+                tuple >> rule;
+
+                info.insert({std::make_pair(name.substr(0, name.find("-@info")), ruleNum), bodyRels});
+                rules.insert({std::make_pair(name.substr(0, name.find("-@info")), ruleNum), rule});
+            }
+        }
+    }
+
+    std::unique_ptr<TreeNode> explain(
+            std::string relName, std::vector<RamDomain> tuple, int ruleNum, int levelNum, size_t depthLimit) {
+        std::stringstream joinedArgs;
+        joinedArgs << join(numsToArgs(relName, tuple), ", ");
+        auto joinedArgsStr = joinedArgs.str();
+
+        // if fact
+        if (levelNum == 0) {
+            return std::unique_ptr<TreeNode>(new LeafNode(relName + "(" + joinedArgsStr + ")"));
+        }
+
+        assert(info.find(std::make_pair(relName, ruleNum)) != info.end() && "invalid rule for tuple");
+
+        // if depth limit exceeded
+        if (depthLimit <= 1) {
+            tuple.push_back(ruleNum);
+            tuple.push_back(levelNum);
+
+            // find if subproof exists already
+            size_t idx = 0;
+            auto it = std::find(subproofs.begin(), subproofs.end(), tuple);
+            if (it != subproofs.end()) {
+                idx = it - subproofs.begin();
+            } else {
+                subproofs.push_back(tuple);
+                idx = subproofs.size() - 1;
+            }
+
+            return std::unique_ptr<TreeNode>(
+                    new LeafNode("subproof " + relName + "(" + std::to_string(idx) + ")"));
+        }
+
+        auto internalNode =
+                new InnerNode(relName + "(" + joinedArgsStr + ")", "(R" + std::to_string(ruleNum) + ")");
+
+        // make return vector pointer
+        std::vector<RamDomain>* ret = new std::vector<RamDomain>();
+        std::vector<bool>* err = new std::vector<bool>();
```

Comment:
Do these need to be pointers? You could just make vectors here and make a pointer to them for prog.executeSubroutine - or change executeSubroutine to use references